### PR TITLE
8314949: linux PPC64 Big Endian: Implementation of Foreign Function & Memory API

### PR DESCRIPTION
**Pull request by TheRealMDoerr**
I've found a way to solve the remaining FFI problem on linux PPC64 Big Endian. Large structs (>8 Bytes) which are passed in registers or on stack require shifting the Bytes in the last slot if the size is not a multiple of 8. This PR adds the required functionality to the Java code.

Please review and provide feedback. There may be better ways to implement it. I just found one which works and makes the tests pass:
```
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/jdk/java/foreign                          88    88     0     0   
```

Note: This PR should be considered as preparation work for AIX which also uses ABIv1.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314949](https://bugs.openjdk.org/browse/JDK-8314949): linux PPC64 Big Endian: Implementation of Foreign Function &amp; Memory API (**Enhancement** - P4)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15417/head:pull/15417` \
`$ git checkout pull/15417`

Update a local copy of the PR: \
`$ git checkout pull/15417` \
`$ git pull https://git.openjdk.org/jdk.git pull/15417/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15417`

View PR using the GUI difftool: \
`$ git pr show -t 15417`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15417.diff">https://git.openjdk.org/jdk/pull/15417.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15417#issuecomment-1691750537)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/ABIv1CallArranger.java
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/ABIv1CallArranger.java:33]: Wouldn't it be more natural for CallArranger to have an abstract method (or even a kind() accessor for the different kinds of ABI supported) and then have these specialized subclasses return the correct kind? It seems to me that setting the `useXYZAbi` flag using an instanceof test is a little dirty.

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java:695]: Given the situation you are facing, perhaps adding the new binding here is unavoidable. Let's wait to hear from @JornVernee. In the meantime, can you point me to a document which explains this behavior? I'm curious and I'd like to know more :-)

**Review by mcimadamore — COMMENTED** [review]
Overall these changes look good - as commented I'd like to learn a bit more of the underlying ABI, to get a sense of whether adding a new binding is ok. But overall it's great to see support for a big-endian ABI - apart from the linker, I am pleased to see that you did not encounter too many issues in the memory-side of the FFM API.

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java:695]: Maybe I'm starting to see it - it's not a special rule, as much as it is a consequence of the endianness. E.g. if you have a struct that is 64 + 32 bytes, you can store the first 64 bytes as a long. Then, there's an issue as we have to fill another long, but we have only 32 bits of value. Is it the problem that if we just copy the value into the long word "as is" it will be stored in the "wrong" 32 bits? So the shift takes care of that, I guess?

Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java:695]: If my assumption above is correct, then maybe another way to solve the problem, would be to, instead of adding a new shift binding, to generalize the VM store binding we have to allow writing a smaller value into a bigger storage, with an offset. Correct?

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java:717]: Why do we assume we can only deal with ints or longs?

File: src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/ABIv1CallArranger.java
Comment (TheRealMDoerr) [reply src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/ABIv1CallArranger.java:33]: I had something like that, but another reviewer didn't like it, either. Originally, I had thought that the v1 and v2 CallArrangers would get more content, but they're still empty. Would it be better to remove these special CallArrangers and distinguish in the base class?

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (TheRealMDoerr) [reply src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java:695]: The ABI says: "An aggregate or union smaller than one doubleword in size is padded so that it appears in the least significant bits of the doubleword. All others are padded, if necessary, at their tail." [https://refspecs.linuxfoundation.org/ELF/ppc64/PPC-elf64abi.html#PARAM-PASS].
I have written examples which pass 9 and 15 Bytes.
In the first case, we need to get 0x0001020304050607 in the first argument and 0x08XXXXXXXXXXXXXX into the second argument (X is "don't care"). Shift amount is 7.
In the second case, we need to get 0x0001020304050607 in the first argument and 0x08090a0b0c0d0eXX into the second argument. Shift amount is 1.
In other words, we need shift amounts between 1 and 7. Stack slots and registers are always 64 bit on PPC64.

**Comment by TheRealMDoerr** [comment]
@mcimadamore: Thanks for your feedback! Jorn and I had resolved the other issues already when we have worked on the linux little endian part. It already contains some ABIv1 code. Note that we already have one big endian platform: s390. But that one doesn't pass structs >8 Bytes in registers.

File: src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/ABIv1CallArranger.java
Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/ABIv1CallArranger.java:33]: It seems to me that what you are doing is similar to what was done for aarch64, which was dealt with using very simple subclasses:
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/linux/LinuxAArch64CallArranger.java
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/macos/MacOsAArch64CallArranger.java
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/windows/WindowsAArch64CallArranger.java

In your case there's less difference, but I think we should follow the same idiom for both.

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java:695]: Got it - I found these representations:

https://refspecs.linuxfoundation.org/ELF/ppc64/PPC-elf64abi-1.7.html#BYTEORDER

Very helpful. So you have e.g. a `short` value (loaded from somewhere) and you have to store it on a double-word. Now, if you just stored it at offset 0, you will write the bits 0-15, which are the "most" significant bits in big-endian representation. So, it's backwards. I believe FFM will take care of endianness, so that the bytes 0-7 and 8-15 will be "swapped" when writing into the double-word (right?) but their base offset (0) is still off, as they should really start at offset 48. Hence the shift.

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java:717]: Ah I see that now - it's done the binding "builder".

File: src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/ABIv1CallArranger.java
Comment (TheRealMDoerr) [reply src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/ABIv1CallArranger.java:33]: Makes sense. I've changed it with the 2nd commit.

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java:695]: Please split this into 2 binding operators: one for ShiftLeft and another for (arithmetic) ShiftRight, rather than mixing the 2 implementations into a single operator.

For the right shift you could then also use a positive shift amount, and avoid the double negation that's needed right now.

Comment (JornVernee) [reply src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java:695]: After looking for a while, I think having the new binding operator is needed. e.g. for stores: we load a 32 bit value from the end of a struct, then the low-order bits of the value needs to be padded with zeros to get a 64 bit register value, leaving the original 32 bit value in the high order bits. This can't be handled by the current cast operator. If we had an int -> long cast conversion the low-order bits would be occupied by the 32 bit value.

**Review by JornVernee — COMMENTED** [review]
Sorry for the delay, I've been on vacation.

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (TheRealMDoerr) [reply src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java]: Done.

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (TheRealMDoerr) [reply src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java:707]: Done.

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (TheRealMDoerr) [reply src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java]: I guess we would need to add additional bindings for that? Is is worth adding more just for a big endian corner case? Or can that be done with the existing ones?

**Comment by TheRealMDoerr** [comment]
> Sorry for the delay, I've been on vacation.

No problem. Hope you had a good time! Thanks for your feedback.

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (JornVernee) [reply src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java]: A `Cast` case for int -> long and long -> int would need to be added. Given the existing setup, that should only be a few lines of code for each. (See e.g. for int -> long https://github.com/openjdk/jdk/compare/master...JornVernee:jdk:I2L). I don't think the cost is that high.

> Is is worth adding more just for a big endian corner case?

I think it's worth it in order to have a cleaner contract for the shift ops, should we want to use them for anything else in the future, but also just to make them easier to understand for future readers.

**Comment by JornVernee** [comment]
@TheRealMDoerr We've been discussing the shifts in order to wrap our heads around it, and we've ended up with this diagram in order to try and visualize what happens:

Let's say we have a struct with 3 ints:

```
struct Foo {
    int x;
    int y;
    int z;
};
```

If this struct is passed as an argument, then the load of the second 'half' of the struct would look like this:

```
offset         : 0 .... 32 ..... 64 ..... 96 .... 128
values         : xxxxxxxx|yyyyyyyy|zzzzzzzz|????????   (can't touch bits 96..128)
Load int       :                  V        +--------+
                                  |                 |
                                  +--------+        |
                                           V        V
In register    :                   ????????|zzzzzzzz   (MSBs are 0)
Shift left     :                   zzzzzzzz|00000000   (LSBs are zero)
Write long     :                  V                 V
Result         : xxxxxxxx|yyyyyyyy|zzzzzzzz|00000000
```

So, the 'Result' is padded at the tail with zeros.

Does that seem right? Does it seem useful to add this diagram as a comment somewhere, for us when we come back to this code a year from now? Thanks

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java]: > I think it's worth it in order to have a cleaner contract for the shift ops, should we want to use them for anything else in the future, but also just to make them easier to understand for future readers.

I agree that having a cleaner contract for the shift binding would prove useful in the long run. If we do that, we can also simplify the binding itself, as it would no longer need an input type?

**Comment by mcimadamore** [comment]
> If this struct is passed as an argument, then the load of the second 'half' of the struct would look like this:

It would perhaps be cleaner if in the MSB/LSB comments we said:

LSBs are zzz...z
LSBs are 000...0

(e.g. avoid to refer to MSBs in the first, since those bytes are not exactly zero, they are the padding bytes)

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (JornVernee) [reply src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java]: > as it would no longer need an input type?

Yes. Then both shift ops would always operate on `long`.

File: src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java
Comment (TheRealMDoerr) [reply src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java]: Ok.

File: src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/abi/Binding.java:425]: This line seems redundant now (`type` is not used below).

**Comment by TheRealMDoerr** [comment]
Hmm. Do you see a good place for such a comment? Maybe it would be better to use a different size for the last chunk. Maybe one or seven Bytes. That's even less straight-forward.

**Comment by JornVernee** [comment]
> Do you see a good place for such a comment?

PPC CallArranger seems like a good place to me. We have a similar explanation comment in the AArch64 CallArranger.

> Maybe it would be better to use a different size for the last chunk. Maybe three or five Bytes. That's even less straight-forward.

Does the size matter that much? It just changes the shift amount right? Could use `short` as type for `z` as well.

**Comment by TheRealMDoerr** [comment]
Thank you! I have added an example to the CallArranger. Please take a look.

**Comment by JornVernee** [comment]
Latest version looks great! I've started a CI job as well. Will approve when that comes back green.

**Comment by TheRealMDoerr** [comment]
@mcimadamore: May I ask you or somebody else from the Panama team to provide a 2nd review? This PR requires Panama knowledge, not really PPC knowledge.

**Comment by mcimadamore** [comment]
> @mcimadamore: May I ask you or somebody else from the Panama team to provide a 2nd review? This PR requires Panama knowledge, not really PPC knowledge.

Sorry - I have already reviewed it - but didn't approve as I was waiting for @JornVernee to chime in. Now he has, and I will add my approval as well.

**Review by mcimadamore — APPROVED** [review]
Looks great - thanks!

**Comment by TheRealMDoerr** [comment]
Thanks for your outstanding support! I'm planning to integrate tomorrow.

**Pull request closed without merging**
